packages/arb-avm-cpp/avm_values/src/value.cpp

```diff
@@ -100,81 +91,83 @@ void marshal_uint256_t(const uint256_t& val, std::vector<unsigned char>& buf) {
     buf.insert(buf.end(), tmpbuf.begin(), tmpbuf.end());
 }
 
-void marshal_value(const value& val,
-                   std::vector<unsigned char>& buf,
-                   const Code& code) {
+void marshal_value(const value& val, std::vector<unsigned char>& buf) {
     if (nonstd::holds_alternative<Tuple>(val)) {
-        nonstd::get<Tuple>(val).marshal(buf, code);
+        nonstd::get<Tuple>(val).marshal(buf);
     } else if (nonstd::holds_alternative<uint256_t>(val)) {
         buf.push_back(NUM);
         marshal_uint256_t(nonstd::get<uint256_t>(val), buf);
     } else if (nonstd::holds_alternative<CodePointStub>(val)) {
-        auto cpr = nonstd::get<CodePointStub>(val);
-        code[cpr].marshal(buf, code);
+        buf.push_back(CODE_POINT_STUB);
+        nonstd::get<CodePointStub>(val).marshal(buf);
     } else if (nonstd::holds_alternative<HashPreImage>(val)) {
+        buf.push_back(HASH_PRE_IMAGE);
         nonstd::get<HashPreImage>(val).marshal(buf);
     }
 }
 
-void marshalStub(const value& val,
-                 std::vector<unsigned char>& buf,
-                 const Code& code) {
-    if (nonstd::holds_alternative<Tuple>(val)) {
-        auto tup = nonstd::get<Tuple>(val);
-        marshalForProof(tup.getHashPreImage(), buf, code);
-    } else {
-        marshalForProof(val, buf, code);
-    }
-}
-
-void marshalForProof(const CodePoint& cp,
+namespace {
+void marshalForProof(const HashPreImage& val,
+                     MarshalLevel,
                      std::vector<unsigned char>& buf,
-                     const Code& code) {
-    buf.push_back(CODEPT);
-    cp.op.marshalForProof(buf, false, code);
-    std::array<unsigned char, 32> hashVal;
-    to_big_endian(cp.nextHash, hashVal.begin());
-    buf.insert(buf.end(), hashVal.begin(), hashVal.end());
+                     const Code&) {
+    buf.push_back(HASH_PRE_IMAGE);
+    val.marshal(buf);
 }
 
-namespace {
-// see similar functionality in tuple.cloneShallow and tuple.marshal
 void marshalForProof(const Tuple& val,
+                     MarshalLevel marshal_level,
                      std::vector<unsigned char>& buf,
                      const Code& code) {
-    buf.push_back(TUPLE + val.tuple_size());
-    for (uint64_t i = 0; i < val.tuple_size(); i++) {
-        auto itemval = val.get_element(i);
-        marshalStub(itemval, buf, code);
+    if (marshal_level == MarshalLevel::STUB) {
+        marshalForProof(val.getHashPreImage(), marshal_level, buf, code);
+    } else {
+        buf.push_back(TUPLE + val.tuple_size());
+        MarshalLevel nested_level = MarshalLevel::STUB;
+        if (marshal_level == MarshalLevel::FULL) {
+            nested_level = MarshalLevel::FULL;
+        }
```

Comment:
level = (marshal_level == MarshalLevel::FULL ? marshal_level : MarshalLevel::STUB) would be more clear

Reply:
also its done several times, cleaner to have one function to compute marshal level

Reply:
I refactored all of these so it's only defined in one place